### PR TITLE
add cascading limits

### PR DESCRIPTION
**Pull request by sleepylemur**
@rmarianski mind taking a look?

### CODE REVIEW

File: proxy/apiaxle-proxy.coffee [hunk 1]
```diff
@@ -292,13 +294,43 @@ class exports.ApiaxleProxy extends AxleApp
...
+    # if keyless, use api's keyless limits
+    if key.id.substring(0, 3) == 'ip-'
+      qps = api.data.keylessQps
+      qpm = api.data.keylessQpm
+      qpd = api.data.keylessQpd
+    else
+      # use key's api specific limits if they exist
+      if key.data.apiLimits
+        try
+          keyApiLimits = JSON.parse key.data.apiLimits
```
Comment (rmarianski) [inline proxy/apiaxle-proxy.coffee:307]: Don't need a `catch` in coffeescript?

File: proxy/apiaxle-proxy.coffee [hunk 2]
```diff
@@ -292,13 +294,43 @@ class exports.ApiaxleProxy extends AxleApp
...
+      if qpd == undefined
+        qpd = key.data.qpd
+
+      # fall back to api default limits
+      if qps == undefined
+        qps = api.data.qps
+      if qpm == undefined
+        qpm = api.data.qpm
+      if qpd == undefined
+        qpd = api.data.qpd
```
Comment (rmarianski) [inline proxy/apiaxle-proxy.coffee:327]: Pasted twice?

File: proxy/apiaxle-proxy.coffee [hunk 3]
```diff
@@ -292,13 +294,43 @@ class exports.ApiaxleProxy extends AxleApp
...
+        qpm = keyApiLimits[api.id].qpm
+        qpd = keyApiLimits[api.id].qpd
+
+      # fall back to key limits
+      if qps == undefined
+        qps = key.data.qps
+      if qpm == undefined
+        qpm = key.data.qpm
+      if qpd == undefined
+        qpd = key.data.qpd
```
Comment (rmarianski) [inline proxy/apiaxle-proxy.coffee:319]: Does coffeescript "do the right thing" here for undefined checks?

File: proxy/apiaxle-proxy.coffee (hunk 1 above)
Comment (sleepylemur) [reply proxy/apiaxle-proxy.coffee:307]: `catch` is optional in coffeescript

File: proxy/apiaxle-proxy.coffee (hunk 2 above)
Comment (sleepylemur) [reply proxy/apiaxle-proxy.coffee:327]: first one is old-style key limits, 2nd one is the default api limits.

File: proxy/apiaxle-proxy.coffee (hunk 3 above)
Comment (sleepylemur) [reply proxy/apiaxle-proxy.coffee:319]: the code works as expected. not sure if it's the best way to do it, but this causes qps, qpm, and qpd to fall through individually.
so first check the apiLimits stored in the key, then check the legacy qps/qpm/qpd stored in the key, then check qps/qpm/qpd stored in the api.

Comment (rmarianski) [reply proxy/apiaxle-proxy.coffee:319]: :white_check_mark:

File: proxy/apiaxle-proxy.coffee (hunk 2 above)
Comment (rmarianski) [reply proxy/apiaxle-proxy.coffee:327]: :white_check_mark:

**Comment by rmarianski** [comment]
:+1:

**Pull request merged**